blockchain/vm/contracts.go

```diff
@@ -825,3 +862,501 @@ func (c *validateSender) validateSender(input []byte, picker types.AccountKeyPic
+	// > Output is an encoding of addition operation result - single G1 point (`128` bytes).
+	if len(input) != 256 {
+		return nil, errBLS12381InvalidInputLength
+	}
+	var err error
+	var p0, p1 *bls12381.G1Affine
+
+	// Decode G1 point p_0
+	if p0, err = decodePointG1(input[:128]); err != nil {
+		return nil, err
+	}
+	// Decode G1 point p_1
+	if p1, err = decodePointG1(input[128:]); err != nil {
+		return nil, err
+	}
+
+	// No need to check the subgroup here, as specified by EIP-2537
+
+	// Compute r = p_0 + p_1
+	p0.Add(p0, p1)
+
+	// Encode the G1 point result into 128 bytes
+	return encodePointG1(p0), nil
+}
+
+// bls12381G1Mul implements EIP-2537 G1Mul precompile.
+type bls12381G1Mul struct{}
+
+// GetRequiredGasAndComputationCost returns the gas required to execute the pre-compiled contract.
+func (c *bls12381G1Mul) GetRequiredGasAndComputationCost(input []byte) (uint64, uint64) {
+	return params.Bls12381G1MulGas, params.Bls12381G1MulComputationCost
+}
+
+func (c *bls12381G1Mul) Run(input []byte, contract *Contract, evm *EVM) ([]byte, error) {
+	// Implements EIP-2537 G1Mul precompile.
+	// > G1 multiplication call expects `160` bytes as an input that is interpreted as byte concatenation of encoding of G1 point (`128` bytes) and encoding of a scalar value (`32` bytes).
+	// > Output is an encoding of multiplication operation result - single G1 point (`128` bytes).
+	if len(input) != 160 {
+		return nil, errBLS12381InvalidInputLength
+	}
+	var err error
+	var p0 *bls12381.G1Affine
+
+	// Decode G1 point
+	if p0, err = decodePointG1(input[:128]); err != nil {
+		return nil, err
+	}
+	// 'point is on curve' check already done,
+	// Here we need to apply subgroup checks.
+	if !p0.IsInSubGroup() {
+		return nil, errBLS12381G1PointSubgroup
+	}
+	// Decode scalar value
+	e := new(big.Int).SetBytes(input[128:])
+
+	// Compute r = e * p_0
+	r := new(bls12381.G1Affine)
+	r.ScalarMultiplication(p0, e)
+
+	// Encode the G1 point into 128 bytes
+	return encodePointG1(r), nil
+}
+
+// bls12381G1MultiExp implements EIP-2537 G1MultiExp precompile.
+type bls12381G1MultiExp struct{}
+
+// GetRequiredGasAndComputationCost returns the gas required to execute the pre-compiled contract.
+func (c *bls12381G1MultiExp) GetRequiredGasAndComputationCost(input []byte) (uint64, uint64) {
+	// Calculate G1 point, scalar value pair length
+	k := len(input) / 160
+	if k == 0 {
+		// Return 0 gas for small input length
+		return 0, 0
+	}
+	// Lookup discount value for G1 point, scalar value pair length
+	var discount uint64
+	if dLen := len(params.Bls12381MultiExpDiscountTable); k < dLen {
+		discount = params.Bls12381MultiExpDiscountTable[k-1]
+	} else {
+		discount = params.Bls12381MultiExpDiscountTable[dLen-1]
+	}
+	// Calculate gas and return the result
+	return (uint64(k) * params.Bls12381G1MulGas * discount) / 1000,
+		710 * (uint64(k) * params.Bls12381G1MulGas * discount) / 1000 / 16
```

Comment:
Because we already have Bls12381G1MulComputationCost = Bls12381G1MulGas * 710 / 16,
please write `uint64(k) * params.Bls12381G1MulComputationCost * discount) / 1000`

Reply:
Fixed https://github.com/kaiachain/kaia/pull/111/commits/66e1d18b02dd1004680a42fe1cb4a5360fa0f567